### ISSUE TITLE
Perfomance degradation in raylet task submission

### ISSUE THREAD

**Issue by simon-mo** [ray-project/ray#5310]
We identified that this commit https://github.com/ray-project/ray/commit/40395acadf907107b5e41bc90e8ba603cdce6203 has caused a performance degradation about (2-3x slower) in terms of raylet task submission. In particular: 

- Regular task submission

|        | mean(ms) | p25(ms) | p50(ms) | p75(ms) |
|--------|----------|---------|---------|---------|
| Before | 0.189    | 0.064   | 0.077   | 0.081   |
| After  | 0.380    | 0.327   | 0.348   | 0.372   |

- Actor task submission

|        | mean(ms) | p25(ms) | p50(ms) | p75(ms) |
|--------|----------|---------|---------|---------|
| Before | 0.093    | 0.063   | 0.064   | 0.098   |
| After  | 0.319    | 0.277   | 0.308   | 0.340   |

Steps to reproduce:
- https://gist.github.com/simon-mo/ae92d1f8e37e696738fe346f53a9e64d

**Comment by pcmoritz** [comment]
I can reproduce this:

before:
```
count    24000.000000
mean         0.183129
std          0.419055
min          0.080441
25%          0.147238
50%          0.151181
75%          0.155902
max         24.122023
```

after:
```
count    24000.000000
mean         0.950430
std          0.437788
min          0.433465
25%          0.634165
50%          0.789399
75%          1.201633
max          7.524983
```

**Comment by raulchen** [comment]
Sorry for introducing this issue. 

When developing #5120, we initially let raylet client send `SubmitTaskRequest`s asynchronously.  However, this makes submitting tasks much faster than before, and triggered a bug. That is, when node manager suddenly receives many requests, a lot of handlers are posted to the asio event loop. And asio won't handle the heartbeat timer callback before finishing all these handlers, and then cause heartbeat timeout.

Then we ended up submitting tasks synchronously. Because, previously, `ClientConnection` was also synchronous (it reads next request from buffer when the previous request is finishes). By the way, we did a local perf test, but didn't see obvious regression. Probably it's related to the environment.

@jiangzihao2009 is working on fixing the heartbeat timer issue and changing submitting tasks back to async at #5296 right now.  And that PR fix this regression.

**Comment by robertnishihara** [comment]
Thanks @raulchen! When do we expect to merge #5296? If it's later than today, maybe the fix should be factored out into a separate PR? We will delay the 0.7.3 release until this is fixed.

**Comment by raulchen** [comment]
We'll try to get that PR done today (July 30 China time).

**Comment by raulchen** [comment]
The root cause turned out to be redis latency, not the post handlers. I created another simpler PR to fix this #5313
